### ISSUE TITLE
Expiration date format

### ISSUE THREAD

**Issue by hangy** [openfoodfacts/robotoff#22]
I noticed that [revision 29](https://de.openfoodfacts.org/produkt/4311501634684/leo-crisp-gut-gunstig?rev=29), robotoff added the expiration date `14/06/2019`. In the [JSON file](https://de.openfoodfacts.org/images/products/431/150/163/4684/11.json) for my uploaded picture, the date is written as `14.06.2019` (dd.mm.yyyy). Clearly, there's some kind of processing going on. In my opinion, the date should either be written in a format in the language of the uploaded picture, or normalized to [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601), so that consumers don't need to play a guessing game about which digit is a day and which is a month. I prefer ISO 8601 for all languages.

**Comment by raphael0202** [comment]
Indeed there is a preprocessing going on, first to check that the candidate is really a date, and to normalize the date format. However only dates of format `%d/%m/%Y` are currently recognized, so `mm.dd.yyyy` dates are not recognized (unless the value for month is valid for day and inversely).

See https://github.com/openfoodfacts/robotoff/blob/master/robotoff/insights/ocr/expiration_date.py for more info.
I like the idea of normalizing to ISO 8601, but the format will be inconsistent between robotoff- and user- annotated products.

**Comment by hangy** [comment]
The regex `full_digits_long` does match `14.06.2019` if I test it: https://regexr.com/4br3a Otherwise, robotoff wouldn't have added `14/06/2019` as the products date.

**Comment by raphael0202** [comment]
Yes indeed, as I said above robotoff matches dates of format `dd.mm.yyyy`

**Comment by hangy** [comment]
Sorry, I must've misread. In that case, robotoff shouldn't replace the separators.

> I like the idea of normalizing to ISO 8601, but the format will be inconsistent between robotoff- and user- annotated products.

I usually do use ISO when editing manually, because it's the only consistent syntax. 😁 But yes, there may be differences. There's an old openfoodfacts-server issue about normalizing the date, but it work on it hasn't been started, yet.

**Comment by hangy** [comment]
> Yes indeed, as I said above robotoff matches dates of format `dd.mm.yyyy`

In https://de.openfoodfacts.org/produkt/4311501619872/harzer-minis-gut-gunstig?rev=20 the expiration date was updated to `06/07/2019` based on the text `06.07.19` in this image: https://de.openfoodfacts.org/images/products/431/150/161/9872/4.jpg The product's main language is German, an `dd/mm/yyyy` is not a known date pattern in Germany.

**Comment by raphael0202** [comment]
I've fixed the normalization issue by normalizing dates to ISO 8601 in 836b4eb82a832498b29f676d917f58ba1c26a13f.
Thanks for the report!

Regarding the other issue you mention, I think we could change the detection pattern given the detected language on the image. If most words returned by the OCR are detected as german -> `dd/mm/yyyy` pattern would not be used.

**Comment by hangy** [comment]
> I've fixed the normalization issue by normalizing dates to ISO 8601 in [836b4eb](https://github.com/openfoodfacts/robotoff/commit/836b4eb82a832498b29f676d917f58ba1c26a13f).

Looks good, thank you!

> Regarding the other issue you mention, I think we could change the detection pattern given the detected language on the image. If most words returned by the OCR are detected as german -> `dd/mm/yyyy` pattern would not be used.

That's probably a good idea. [Wikipedia](https://en.wikipedia.org/w/index.php?title=Date_format_by_country&oldid=925020095) lists several interesting date patterns that could be used for parsing.

**Comment by PixelCode01** [comment]
I'd like to work on this issue.
I looked at the commit history, and the ISO normalization fix is already done. However, the locale-aware parsing of ambiguous expiration dates appears to be an open issue. 
My proposed changes are:
* Improve language-aware parsing of ambiguous numeric dates (like 06.07.19 cases),
* Preserve the normalized output consistent with the existing behavior,
* Provide targeted tests for German/French/English patterns and ambiguity cases.
If the above scope is correct, I'd like to create a small PR first for the tests, followed by the implementation.
